runtime/lua/vim/shared.lua

```diff
@@ -300,6 +300,29 @@ function vim.tbl_islist(t)
   end
 end
 
+--- Returns the first element from the table that matches the second argument.
+---
+--@param t table
+--@param o Any object. If f is callable, apply the element to f, otherwise compare the element to f.
+function vim.tbl_find(t, o) end -- luacheck: no unused
+vim.tbl_find = (function()
+  local function _apply(v, o)
+    if vim.is_callable(o) then
+      return o(v)
+    else
+      return v == o
+    end
+  end
+
+  return function(t, o)
+    vim.validate{t={t, 't'}}
+    for _, v in pairs(t) do
+      if _apply(v, o) then return v end
+    end
+    return nil
+  end
+end)()
```

Comment:
I think we don't need this pattern here. And we can avoid `_apply`  checking  if o is callable for each comparison. Something like this (untested):

```
function vim.tbl_find(t, o)
  vim.validate{t={t, 't'}}
  local cmp = vim.is_callable(o) and o or function(v) o == v end
  for _, v in pairs(t) do
    if cmp(v) then return v end
  end
  return nil
end
```